### ISSUE TITLE
Please guide how to send email using gmail api via gsuit service account

### ISSUE THREAD

**Issue by meepeek** [googleapis/google-api-nodejs-client#2418]
My goal is for my nodejs app to send mail to notify me if there were unhandled code exception on the server.

The code below is what I use for testing, which I modified gmail api quickstart code to use keyFile instead of oauth. However, I stuck with the error "The API returned an error: Error: Invalid Credentials".

I use this auth code with spreadsheet api before and it was success. I also did 
Domain-wide Delegation and add the app to Google Admin access control.

Now I'm stuck and cannot find any nodejs example for gmail. Please help.

```
import googleapis from 'googleapis'

const {google} = googleapis

const auth = new google.auth.GoogleAuth({
    keyFile: './credentials/gmail-key.json',
    scopes: ['https://www.googleapis.com/auth/gmail.send,email,profile']
})

const gmail = google.gmail({version: 'v1', auth})

// console.log(gmail.users.messages.send)

listLabels(auth)

function listLabels(auth) {
    const gmail = google.gmail({version: 'v1', auth});
    gmail.users.labels.list({
// I use service account email
      userId: 'SERVICE_ACCOUNT_NAME@APP_NAME.iam.gserviceaccount.com',
    }, (err, res) => {
      if (err) return console.log('The API returned an error: ' + err);
      const labels = res.data.labels;
      if (labels.length) {
        console.log('Labels:');
        labels.forEach((label) => {
          console.log(`- ${label.name}`);
        });
      } else {
        console.log('No labels found.');
      }
    });
  }
```

**Comment by meepeek** [comment]
PS. If I use the same scope as in quickstart, 'https://www.googleapis.com/auth/gmail.readonly', it will print "The API returned an error: Error: Precondition check failed."

**Comment by JustinBeckwith** [comment]
@sqrrrl we get fairly frequent requests to help figure out GSuite + Service Account issues.  Do you know if there's a canonical guide somewhere out there on the topic?

**Comment by sqrrrl** [comment]
There's not, but something our tech writers are looking at improving.

Most Workspace APIs expect to be called as an end-user, not a service account. Support for services accounts is the exception, not the rule, and there's only a handful of APIs where it's appropriate (e.g. Drive, though still discouraged.) Gmail and Calendar specifically do *not* allow it.

They all support domain-wide delegation where a service account is used to impersonate an end-user. But that's not what this code example is doing. It's using the service account identity itself which isn't allowed here. To do delegation, the credentials need to be scoped to the target user by setting the sub claim in the token. Looks like for this client that means creating the JWT client directly (https://github.com/googleapis/google-auth-library-nodejs/blob/master/src/auth/jwtclient.ts) and setting the subject arg to the user's email address. Some of the other libraries have convenience methods for doing this (similar to createScoped(scopes) -- createDelegated(user)) and that could be useful to add here.

The other change would be using the user email (or the 'me' alias which just means whoever the effective user is) for the userId parameter in the API request.

To summarize:
* Workspace APIs should (almost) always be called as an end-user, not a service account
* Service accounts are useful for admin-managed apps that need to impersonate users in a domain without their explicit consent (effective user for the credential is still an end-user though, consistent with the first point.)
* Node.js client could make that a little easier with convenience methods to get a delegated credential

**Comment by Akxe** [comment]
I too find the documentation around sending emails with the service account.

```ts
import { google } from 'googleapis';

class MailProvider {
	private gmail = google.gmail({
		version: 'v1',
		auth: new google.auth.GoogleAuth({
			keyFile: '../../assets/secrets/google.json',
			scopes: [
				'https://www.googleapis.com/auth/gmail.send',
				'https://www.googleapis.com/auth/gmail.readonly',
			],
		}),
	});

	async sendMail(sendAs: string, sendTo: string) {
		return this.gmail.users.messages.send({
			userId: sendAs,
			media: {
				mimeType: 'placeholder-value',
				body: 'placeholder-value',
			},
		});
	}
}

new MailProvider().sendMail('noreply@example.com', 'akxe@seznam.cz'); // noreply@example.com is email from domain
```

It is very hard to find what should I put to the `media.mimeType` property. The example has `placeholder-value`, not really useful. Also, where do I put the receiver of the message? Stack overflow only mentions the `media.raw` property, but everyone is composing the raw manually, I would like not to do that if possible...

**Comment by Tallyrald** [comment]
I found a way to make this work in a relatively painless way so I'll share in case someone needs it later.
As you know, the docs are kind of useless for us nodejs folks so I'll put some explaining here too.

First of all, create the service account via Cloud Console or other methods & give it domain-wide delegation. No need to give any other roles or permissions as the GMail API is not exactly a GCloud app.
Then head over to the Google Workspaces Admin website (<https://admin.google.com>) and under security -> API setting (or something similarly named) put the delegation ClientId into the allowed apps list. This is where you set the appropriate scopes too.

Now consider the following (tl;dr at bottom):

```typescript
// Warning lots of comments ahead
import googleapis from 'googleapis';
import { encode } from 'js-base64'; // btoa can probably be used too

const { google } = googleapis;

/** Sends a test email (to & from) a test email account */
public static async SendTestEmail() {
    // Create an auth solution. This could be any authentication method that the googleapis package allows
    const authClient = new google.auth.JWT({
        keyFile: 'path/to/keyFile.json', // Service account key file
        scopes: ['https://mail.google.com/'], // Very important to use appropriate scopes. This one gives full access (only if you gave this to the service account too)
        subject: 'email-address-to-impersonate@your-domain.com', // This is an email address that exists in your Google Workspaces account. The app will use this as 'me' during later execution.
    });

    // I'm not sure if this is necessary, but it visualizes that you are now logged in.
    await authClient.authorize();

    // Let's create a Gmail client via the googleapis package
    const gmail = new gmail_v1.Gmail({ auth: authClient });

    // The names are a bit confusing, but the important part is the users.messages.send is responsible for executing the send operation.
    // This means a single email, not batch sending
    const result = await gmail.users.messages.send({
        auth: authClient, // Pass the auth object created above
        requestBody: {
            // I'm fairly certain that the raw property is the most sure-fire way of telling the API what you want to send
            // This is actually a whole email, not just the body, see below
            raw: this.MakeEmail('email-to-send-to@target-domain.com', 'email-to-send-from@your-domain.com', 'Subject string', 'Email body string'),
        },
        userId: 'me', // Using me will set the authenticated user (in the auth object) as the requester
    });

    return result;
}

/** Creates a very basic email structure
 * @param to Email address to send to
 * @param from Email address to put as sender
 * @param subject The subject of the email [warn: encoding, see comment]
 * @param message The body of the email
 */
private static MakeEmail(to: string, from: string, subject: string, message: string) {
    // OK, so here's the magic
    // The array is used only to make this easier to understand, everything is concatenated at the end
    // Set the headers first, then the recipient(s), sender & subject, and finally the message itself
    const str = [
        'Content-Type: text/plain; charset="UTF-8"\n', // Setting the content type as UTF-8 makes sure that the body is interpreted as such by Google
        'to: ', to,'\n',
        'from: ', from,'\n',
        // Here's the trick: by telling the interpreter that the string is base64 encoded UTF-8 string, you can send non-7bit-ASCII characters in the subject
        // I'm not sure why this is so not intuitive (probably historical/compatibility reasons),
        // but you need to make sure the encoding of the file, the server environment & everything else matches what you specify here
        'subject: =?utf-8?B?', encode(subject, true),'?=\n\n', // Encoding is base64 with URL safe settings - just in case you want a URL in the subject (pls no, doesn't make sense)
        message, // The message body can be whatever you want. Parse templates, write simple text, do HTML magic or whatever you like - just use the correct content type header
    ].join('');

    const encodedMail = encode(str, true); // Base64 encode using URL safe settings

    return encodedMail;
}
```

tl;dr:

- Authenticate an account email address to impersonate
- Compose email body however you want (and make sure the content type header reflects what you've put there)
- Concatenate email headers, recipient(s), sender, subject and body into a single string
- Base64 encode the whole thing
- Send it to the API

I haven't used attachments yet. If or when I do, I'll probably update this.

**Comment by radziszXTRF** [comment]
@Tallyrald - great hint thanks
In admin panel (https://admin.google.com) one should go to "API controls" and then select "Domain wide delegation" (below) not "App access control" above, both support provisioning of clientId...

**Comment by djkramnik** [comment]
This use case is probably one of the most common and literally the only reason why I signed up for gsuite (or google workspaces).  Why you don't have a simple guide to this is beyond me.  When I look at the admin console for google workspace I see hundreds of little icons / links / etc that I have zero interest in.  I just want to programatically send emails from my business gmail account, and I don't want to grok whatever half baked, soon to be deprecated cloud architecture you guys are wrapping it in.

**Comment by andris9** [comment]
As people often end up in this issue regarding Gmail Service Accounts, then you can find a tutorial for generating the Service Account key file [here](https://docs.emailengine.app/gmail-oauth-service-accounts/) (the tutorial is not for Nodemailer but EmailEngine, another project I maintain, but the end result is the same, you  need that JSON formatted key file to extract client ID and service key values).

**Comment by gg-martins091** [comment]
The php sdk has a way to set the subject of a client used by a service pretty easliy.

It would be great to have something like this:
```js
const jwtClient = new google.auth.JWT({...})
const service = google.calendar({auth: jwtClient, ....})

service.setSubject('newsubject@email.com')
```

In my use case I'm always using a service account on the Workspaces api and almost always impersonating a user through domain wide delegation, so it would be really helpful to have something like that.

**Comment by kmasterycsl** [comment]
Thanks @andris9 Tallyrald. It's working!

**Comment by Tallyrald** [comment]
Hello everyone! It's been a little more than 5 years since my [comment](https://github.com/googleapis/google-api-nodejs-client/issues/2418#issuecomment-783494711). A lot happened in this time and I was glad that I could help others with my approach. I am not so glad that documentation around this is still sparse.
Now I'm back with another version, because recently Google started throwing `invalid grant: invalid issuer: null` errors for me and I spent another ungodly amount of time scouring the internet and all available options without any success in finding an explanation.

In my case the origin of the error turned out to be the google.auth.JWT function. So I replaced it with google.auth.GoogleAuth.
I also made enhancements, removing the need for the 'js-base64' library, introducing MailComposer as suggested by @kmasterycsl  as well as adding attachment support.

Here is my current code (I had to adapt it for public viewing, hopefully I didn't introduce any errors in the process):
```typescript
import { google } from 'googleapis';
import MailComposer from 'nodemailer/lib/mail-composer';

/** Sends a test email (to & from) a test email account
 * [ATTENTION] This returns a Promise. In my previous comment it returned the result object.
 */
public static SendTestEmail() {
    // Create an auth solution. This used to be the JWT method, but I changed it since I was running into invalid_grant issues.
    const authClient = new google.auth.GoogleAuth({
        clientOptions: {
            subject: 'person_to_impersonate@domain.com', // This is an email address that exists in your Google Workspaces account. The app will use this as 'me' during later execution.
        },
        keyFile: 'your-service-account.json', // Service account key file
        scopes: ['https://mail.google.com/'], // Very important to use appropriate scopes. This one gives full access (only if you gave this to the service account too)
    });

    const gmail = google.gmail({
        auth: authClient,
        version: 'v1',
    });

    const attachments = [{
        contentType: 'binary/octet-stream', // MIME-type
        fileName: 'a-file-name.txt',
        path: 'fully-qualified-path-to-file'
    }];

    // The names are a bit confusing, but the important part is the users.messages.send is responsible for executing the send operation.
    // This means a single email, not batch sending
    return gmail.users.messages.send({
        requestBody: {
            // I'm fairly certain that the raw property is the most sure-fire way of telling the API what you want to send
            raw: await MakeEmail('recipient@domain.com', 'sender@domain.com', 'bcc@domain.com', 'cc@domain.com', 'Email Subject', 'email body', attachments),
        },
        userId: 'me',
    });
}

// Feel free to change the signature of the following function to an object, I just used simple params to show what these are.

/** Creates the email structure.
 * @param to Email address to send to
 * @param from Email address to put as sender
 * @param bcc BCC field
 * @param cc CC field
 * @param subject The subject of the email
 * @param message The body of the email
 * @param attachments Attachment list data
 */
private static async MakeEmail(
    to: string,
    from: string,
    bcc: string,
    cc: string,
    subject: string,
    message: string,
    attachments?: { contentType: string, fileName: string, path: string }[],
) {
    // Type provided by NodeMailer
    const options: SendMailOptions = {
        attachments,
        bcc,
        cc,
        from,
        html: message,
        subject,
        textEncoding: 'base64',
        to,
    };
    // We use MailComposer from NodeMailer, see https://nodemailer.com/extras/mailcomposer
    const mailComposer = new MailComposer(options);
    // This is a buffer containing a base64 encoded string that is our full email
    const compiledMessage = await mailComposer.compile().build();

    // Finally we convert from base64 to base64url. I guess this could be done differently as well since Node seems to support base64url natively,
    // but I found that being explicit with the conversion details works better in my brain.
    // During my testing I found that the Gmail API requires this encoding to be correct
    return Buffer.from(compiledMessage).toString('base64').replaceAll('+', '-').replaceAll('/', '_').replace(/=+$/, '');
}
```

Cheers to all, take care!
